File: 백준/Silver/27737. 버섯 농장/버섯 농장.java
import java.io.*;
import java.util.*;

public class Main {

    static int map[][], N, M, K, area[];
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        N = Integer.parseInt(st.nextToken()); // map 크기
        M = Integer.parseInt(st.nextToken()); // 포자 수
        K = Integer.parseInt(st.nextToken()); // 가능한 연결 칸 수

        map = new int[N][N];
        area = new int[N*N + 1];

        for(int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            for(int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(st.nextToken()) * -1;
            }
        }

        int cnt = 0;
        for(int i = 0; i < N; i++) {
            for(int j = 0; j < N; j++) {
                if(map[i][j] == 0) {
                    dfs(i, j, ++cnt);
                }
            }
        }

        int needs = 0;
        for(int i = 1; i < area.length; i++) {
            if(area[i] > 0) {
                needs += (area[i] % K == 0 ? area[i] / K : area[i] / K + 1);
            }
        }

        if(cnt > 0 && M >= needs) {
            System.out.printf("%s\n%d", "POSSIBLE", M - needs);
        } else {
            System.out.println("IMPOSSIBLE");
        }
    }

    static int dr[] = {-1, 1, 0, 0};
    static int dc[] = {0, 0, -1, 1};
    private static void dfs(int r, int c, int idx) {
        map[r][c] = idx;
        area[idx]++;

        for(int d = 0; d < 4; d++) {
            int nr = r + dr[d];
            int nc = c + dc[d];

            if(nr < 0 || nc < 0 || nr >= N || nc >= N)
                continue;
            if(map[nr][nc] == 0) {
                dfs(nr, nc, idx);
            }
        }
    }
}

File: 백준/Silver/27737. 버섯 농장/버섯 농장.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().trim().split('\n').map((el) => el.split(' ').map(Number));
let N = input[0][0]; // map 크기
let M = input[0][1]; // 포자 수
let K = input[0][2]; // 연결 수
let map = input.slice(1);
let area = Array(Math.floor(N * N / 2) + 2).fill(0); // 인접한 구역별 칸 수 관리
let visited = Array.from({length: N}, () => Array(N).fill(false));

class Spore {
    constructor(r, c) {
        this.r = r;
        this.c = c;
    }
}
const dr = [-1, 1, 0, 0];
const dc = [0, 0, -1, 1];

// 자원 재정의
map.forEach((row, i) => {
   row.forEach((v, j) => {
       if(v === 1) {
           map[i][j] = -1;
       }
    });
});

let cnt = 0;
map.forEach((row, i) => {
    row.forEach((v, j) => {
        if(v === 0 && !visited[i][j]) {
            bfs(i, j, ++cnt);
        }
    });
});

let needs = 0; // 필요한 포자 수
for(let i = 1; i < area.length; i++) {
    if(area[i] > 0) {
        needs += (area[i] % K === 0 ? Math.floor(area[i] / K) : Math.floor(area[i] / K) + 1);
    }
}

if(cnt > 0 && M >= needs) {
    // cnt > 0: 최소 하나라도 포자를 사용해야 함 조건 체크
    // M >= needs: 가지고 있는 포자 수 내에 필요한 포자 수 가능 체크
    console.log('POSSIBLE');
    console.log(M - needs);
} else {
    console.log('IMPOSSIBLE');
}

function bfs(r, c, idx) {
    let q = [];
    q.push(new Spore(r, c));
    visited[r][c] = true;

    while(q.length > 0) {
        let curr = q.shift();
        let cr = curr.r;
        let cc = curr.c;
        map[cr][cc] = idx;
        area[idx]++;

        for(let d = 0; d < 4; d++) {
            let nr = cr + dr[d];
            let nc = cc + dc[d];

            if(nr < 0 || nc < 0 || nr >= N || nc >= N)
                continue;
            if(!visited[nr][nc] && map[nr][nc] === 0) { // 0이 아닌 구역은 불가지역이거나 탐색된(인덱스 표기)구역
                q.push(new Spore(nr, nc));
                visited[nr][nc] = true;
            }
        }
    }
}